### PR TITLE
Improve quick measurement overlays across all possible arrangement scenarios

### PR DESCRIPTION
**Pull request by singhutsav5502**
Adds further code implementation for #1896. This PR adds code that handles all 9 cases as described within the issue, support for individually rotated objects still needs to be added.

### CODE REVIEW

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
!build

**Comment by singhutsav5502** [comment]
The Clippy build fail seems unrelated to the code i have changed.

**Comment by Keavon** [comment]
Yeah, that randomly began failing recently. It's because PR from forks can't access our repo's secret key used by the bot to write the results comment. But it wasn't an issue before. I'm still working on figuring out a strategy to deal with that, such as maybe getting it to send the comment text to a separate workflow that is allowed to access the secret key in order to make the API call to make the comment. (If you happen to have any experience in these matters, feel free to give suggestions, but that's the context here.)

**Comment by Keavon** [comment]
Nice, comprehensive work on this! Some quick QA feedback:

All the cases presented in the issue are working as expected, except for the first two scenarios. Could you please take a closer look at those two? Please pay attention, specifically, to the dashed lines, where the number labels go, and for the first scenario also how the lines connect to the corners of the opposing corner.

Thank you!

**Comment by singhutsav5502** [comment]
Thanks! unfortunately i have no experience with clippy, Will fix the first 2 cases ASAP.

**Comment by singhutsav5502** [comment]
I have pushed the changes for the first 2 cases @Keavon

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
Ok, nice! A few notes:
- The dashed line should please be 2px of blue then 2px of blank
- It seems like, once activated, the dash begins infecting a bunch of the other overlays like the transform cage itself; whatever engineering choice was made to cause this to happen seems to be stateful? Assuming it's stateful, please try to pick an alternative state-free approach so this doesn't have any possibility of happening.

**Comment by singhutsav5502** [comment]
I see, i probably overlooked the issue. I have been using the `dashed_line` method of the `overlay_context` to render the dashed lines. I will try to come up with a solution that uses the normal line method instead so that the transform cage doesn't get affected.

**Comment by Keavon** [comment]
Oh I see, it's the JS drawing API that's keeping the state. I believe you just need to add this to the very end of the `pub fn dashed_line()` function to reset it:

```rs
self.render_context
	.set_line_dash(&JsValue::from(js_sys::Array::new()))
	.map_err(|error| log::warn!("Error drawing dashed line: {:?}", error))
	.ok();
```

Since that, I believe, should satisfy [MDN's note](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash):

> Note: To return to using solid lines, set the line dash list to an empty array.

**Comment by singhutsav5502** [comment]
hi @Keavon i have pushed the needed changes.

**Comment by Keavon** [comment]
!build

File: editor/src/messages/portfolio/document/overlays/utility_types.rs
Comment (Keavon) [inline editor/src/messages/portfolio/document/overlays/utility_types.rs:100]: Please replicate this comment in the function above.

File: editor/src/messages/portfolio/document/overlays/utility_types.rs
Comment (Keavon) [inline editor/src/messages/portfolio/document/overlays/utility_types.rs:81]: What's your reasoning for making two similar functions? I think we could reduce the lines of code by just making the one function more customizable to encompass all the requirements.

**Comment by Keavon** [comment]
Please make the dash pattern 2 pixels of blue, 2 pixels of blank. I think right now it's 3 pixels of blue, 1 pixel of blank?

File: editor/src/messages/portfolio/document/overlays/utility_types.rs
Comment (singhutsav5502) [reply editor/src/messages/portfolio/document/overlays/utility_types.rs:83]: The pre-existing function seems to be incorrectly defined the `dash_width` is being used as the 2nd parameter of the array which effectively makes it the length of the `gap` between two dashes of the line. and the actual dash width (the 1st parameter) seems to be hard set at `1` . 
And this function seems to be used at multiple other places, so i decided to make an extra function that exposes both gap_width and dash_width as individual parameters. ( since a 2px dash was needed to be followed with  a 2px blank space ).

Comment (Keavon) [reply editor/src/messages/portfolio/document/overlays/utility_types.rs:83]: I see, thanks for the explanation. Can you please refactor it to use one function, preserving the appearance of existing usages (where the width is `gap` and `1` functionality you described)?

File: editor/src/messages/tool/common_functionality/measure.rs
Comment (singhutsav5502) [inline editor/src/messages/tool/common_functionality/measure.rs:42]: based on the function description and the input parameters the dash length and the gap length should both be set to 2px. I will look into this.

Comment (singhutsav5502) [reply editor/src/messages/tool/common_functionality/measure.rs:42]: on actual render it does seem to be 3pixels of blue and 1 pixel of blank. I have probably overlooked something in the function definition. will fix ASAP.

Comment (Keavon) [reply editor/src/messages/tool/common_functionality/measure.rs:42]: Yeah, basically, I'd say it's best if you try to combine these functions into one, and then figure out how to make all existing and new usages display their intended dash spacing so the end result is a function that has parameters that behave expectedly but without changing the actual behavior of other usages. Please make note of each existing dash usage and test to ensure it looks identical afterwards compared to https://dev.graphite.rs (the latest master build). It would also be helpful if you could list those locations in a comment here so I can double check them for you.

File: editor/src/messages/portfolio/document/overlays/utility_types.rs
Comment (singhutsav5502) [reply editor/src/messages/portfolio/document/overlays/utility_types.rs:83]: sure!

File: editor/src/messages/tool/common_functionality/measure.rs
Comment (singhutsav5502) [reply editor/src/messages/tool/common_functionality/measure.rs]: -  the `dashed_line` method was being used in the `grid_overlays.rs` file.

-  the `measure.rs` file ( the one currently being edited ). 

Other than that there doesn't seem to be a direct reference to the method in the entire code base. The `line` method leverages the `dashed_line` method indirectly. 
the `line` method is being directly used by the following: ( i am not yet sure if they indirectly leverage dashed_line in some manner, testing would need to be done. )

- `snapping.rs`
- `pen_tool.rs`
- `utility_functions.rs` for bezier functions.

I have refactored the original function to work properly ( as it did before ) for cases where `gap_width` needs to be set to 1.
I have also set the `gap_width` parameter to None in the appropriate places (`grid_overlays.rs` and the `line` method implementation).

Comment (Keavon) [reply editor/src/messages/tool/common_functionality/measure.rs]: Thanks! Does this also solve the 2x2 dashes appearing as 3x1 dashes? I can do another build once you've resolved that too.

Comment (singhutsav5502) [reply editor/src/messages/tool/common_functionality/measure.rs]: From the things i have tested on my device so far the following seem to be working as expected: Grids, Spline Tool (dash parameters ), Line Tool( dash parameters )  and the Pen Tool ( dash parameters ). Not sure if i missed something but do look into it.

**Comment by Keavon** [comment]
!build

File: editor/src/messages/tool/common_functionality/measure.rs
Comment (singhutsav5502) [reply editor/src/messages/tool/common_functionality/measure.rs]: > Thanks! Does this also solve the 2x2 dashes appearing as 3x1 dashes? I can do another build once you've resolved that too.

It seems to be an issue that was already resolved in the pre existing function with how `1` was being subtracted from the dash_width, The dash width and gap width should work properly they looked to be equidistant on my local build.

**Comment by Keavon** [comment]
With this build, it's 2px blue 1px blank. Can you please adjust accordingly?

**Comment by singhutsav5502** [comment]
> With this build, it's 2px blue 1px blank. Can you please adjust accordingly?
Hi, does the image below seem okay? if so, i will push the code.
image

**Comment by Keavon** [comment]
Hard to tell from the screenshot but probably!

**Comment by singhutsav5502** [comment]
pushed the new code.

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
The dash spacing is perfect now, thanks!

**Comment by Keavon** [comment]
So my only remaining comment about this is the text alignment, however I think that's best to save that for a later PR. Are you interested in working on that after we merge this?

**Comment by singhutsav5502** [comment]
> So my only remaining comment about this is the text alignment, however I think that's best to save that for a later PR. Are you interested in working on that after we merge this?

Sure! Though I think I would like to solve a few more smaller issues from the task list before i come back to this code again.

**Comment by Keavon** [comment]
Sure, sounds good. Also the TODO you left about the rotation will also need to be solved eventually too, but it's also best to wait for a later PR. Feel free to work on these or other issues in any order you prefer! Ping me on Discord and we can also help you pick out good ones, but I think you're at a skill level where you can handle most of the issues without too much challenge (as long as it's not blocked by anything else at the moment which would be context only we may have in our knowledge).

**Comment by singhutsav5502** [comment]
Absolutely. Thanks!
I will try and be a lot more active going forward since I am currently a uni student and plan to come up with a gsoc proposal as well. 👍

**Comment by Keavon** [comment]
Sweet, we'll be glad to have your work over the coming months and working together on your GSoC proposal when the applications open (of course, contingent on our org getting accepted which technically hasn't happened yet because we org applications haven't opened yet).

**Comment by singhutsav5502** [comment]
Hi, is there any actionable left from my side which might be an hinderance for merging this?

**Comment by Keavon** [comment]
I just have to do a code review. I pushed a quick commit with some nit picks (feel free to read over the diff for learning purposes about our stylistic preferences). I need to do a detailed code read-through tomorrow. If you're able to improve the code by means of reducing code duplication, that's my current concern for areas that are most likely to require change requests from me, since I feel like there's a sizable amount of similarity in `measure.rs` within all those similar functions.

**Comment by singhutsav5502** [comment]
> I just have to do a code review. I pushed a quick commit with some nit picks (feel free to read over the diff for learning purposes about our stylistic preferences). I need to do a detailed code read-through tomorrow. If you're able to improve the code by means of reducing code duplication, that's my current concern for areas that are most likely to require change requests from me, since I feel like there's a sizable amount of similarity in `measure.rs` within all those similar functions.


Sure, the biggest part of duplicated code seems to be the call for the `line` function. I can probably wrap it into a smaller function that exposes the only 2 changing variables in each call ( rest of the attributes remain same across all function calls in the `measure.rs` file) , for this specific use case but i am not sure if that would be considered redundant or not.

**Comment by Keavon** [comment]
Sounds like a good change. Any code that does the same thing in multiple places should be consolidated, because that helps reduce the line count of your PR (less code for maintainers to read in that file and thus quicker to understand) and more maintainable by avoiding the possibility of changing one place but forgetting to replicate identical changes in related places. Apply that philosophy everywhere, since we aim for minimal lines of code additions and avoiding duplication wherever it can sensibly be achieved.

**Comment by singhutsav5502** [comment]
> Sounds like a good change. Any code that does the same thing in multiple places should be consolidated, because that helps reduce the line count of your PR (less code for maintainers to read in that file and thus quicker to understand) and more maintainable by avoiding the possibility of changing one place but forgetting to replicate identical changes in related places. Apply that philosophy everywhere, since we aim for minimal lines of code additions and avoiding duplication wherever it can sensibly be achieved.

Hi @Keavon, just pushed a code refactor it reduces the code length quite a bit.

**Comment by Keavon** [comment]
!build

**Review by Keavon — APPROVED** [review]

**Pull request merged**
